src/Cmfd.cpp

```diff
@@ -1788,3 +1822,58 @@ void Cmfd::updateBoundaryFlux(Track** tracks, FP_PRECISION* boundary_flux,
     }
   }
 }
+
+
+/**
+ * @brief Zero the surface currents for each mesh cell and energy group.
+ */
+void Cmfd::zeroSurfaceCurrents() {
+
+  #pragma omp parallel for schedule(guided)
+  for (int r=0; r < _num_x*_num_y; r++) {
+    for (int s=0; s < 8; s++) {
+      for (int e=0; e < _num_moc_groups; e++)
+        _surface_currents(r*8+s,e) = 0.0;
+    }
+  }
+
+  return;
+}
+
+
+/**
+ * @brief 
+ * @param curr_segment the current Track segment
+ * @param current the current to tally
+ * @param fwd boolean indicating direction of integration along segment
+ * @param e the MOC energy group index
+ */
+void Cmfd::tallySurfaceCurrent(segment* curr_segment, 
+                               FP_PRECISION current, bool fwd, int e) {
+
+  if (curr_segment->_cmfd_surface_fwd != -1 && fwd){
```

Comment:
I've updated the code to loop over angles and energy groups within ``Cmfd::tallySurfaceCurrent(...)`` rather than in the ``Solver::tallySurfaceCurrent(...)`` routine. This should be much more computationally efficient than before, and hide more of the CMFD-specific computations from the MOC solver.